File: app/src/main/java/org/michaelbel/moviemade/presentation/features/main/MainPresenter.java
package org.michaelbel.moviemade.presentation.features.main;

import org.jetbrains.annotations.NotNull;
import org.michaelbel.moviemade.core.entity.Movie;
import org.michaelbel.moviemade.core.utils.EmptyViewMode;
import org.michaelbel.moviemade.core.utils.NetworkUtil;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.disposables.CompositeDisposable;

public class MainPresenter implements MainContract.Presenter {

    private int page;
    private MainContract.View view;
    private MainContract.Repository repository;
    private CompositeDisposable disposables = new CompositeDisposable();

    public MainPresenter(MainRepository repository) {
        this.repository = repository;
    }

    @Override
    public void attach(@NotNull MainContract.View view) {
        this.view = view;
    }

    @Override
    public void getNowPlaying() {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) {
            view.setError(EmptyViewMode.MODE_NO_CONNECTION);
            return;
        }

        page = 1;
        disposables.add(repository.getNowPlaying(page)
            .subscribe(response -> {
                List<Movie> results = new ArrayList<>(response.getMovies());
                if (results.isEmpty()) {
                    view.setError(EmptyViewMode.MODE_NO_MOVIES);
                    return;
                }
                view.setContent(results);
            }, e -> view.setError(EmptyViewMode.MODE_NO_MOVIES)));
    }

    @Override
    public void getNowPlayingNext() {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) return;

        page++;
        disposables.add(repository.getNowPlaying(page).subscribe(response -> view.setContent(response.getMovies())));
    }

    @Override
    public void getTopRated() {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) {
            view.setError(EmptyViewMode.MODE_NO_CONNECTION);
            return;
        }

        page = 1;
        disposables.add(repository.getTopRated(page)
            .subscribe(response -> {
                List<Movie> results = new ArrayList<>(response.getMovies());
                if (results.isEmpty()) {
                    view.setError(EmptyViewMode.MODE_NO_MOVIES);
                    return;
                }
                view.setContent(results);
            }, e -> view.setError(EmptyViewMode.MODE_NO_MOVIES)));
    }

    @Override
    public void getTopRatedNext() {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) return;

        page++;
        disposables.add(repository.getTopRated(page).subscribe(response -> view.setContent(response.getMovies())));
    }

    @Override
    public void getUpcoming() {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) {
            view.setError(EmptyViewMode.MODE_NO_CONNECTION);
            return;
        }

        page = 1;
        disposables.add(repository.getUpcoming(page)
            .subscribe(response -> {
                List<Movie> results = new ArrayList<>(response.getMovies());
                if (results.isEmpty()) {
                    view.setError(EmptyViewMode.MODE_NO_MOVIES);
                    return;
                }
                view.setContent(results);
            }, e -> view.setError(EmptyViewMode.MODE_NO_MOVIES)));
    }

    @Override
    public void getUpcomingNext() {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) return;

        page++;
        disposables.add(repository.getUpcoming(page).subscribe(response -> view.setContent(response.getMovies())));
    }

    @Override
    public void destroy() {
        disposables.dispose();
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/features/main/MainPresenter.kt
package org.michaelbel.moviemade.presentation.features.main

import org.michaelbel.moviemade.core.errors.EmptyViewMode
import org.michaelbel.moviemade.core.net.NetworkUtil
import org.michaelbel.moviemade.presentation.base.Presenter
import retrofit2.HttpException

class MainPresenter(
        private val repository: MainContract.Repository
): Presenter(), MainContract.Presenter {

    private var page: Int = 0
    private lateinit var view: MainContract.View

    override fun attach(view: MainContract.View) {
        this.view = view
    }

    override fun movies(movieId: Int, list: String) {
        if (!NetworkUtil.isNetworkConnected()) {
            view.error(EmptyViewMode.MODE_NO_CONNECTION)
            return
        }

        page = 1
        disposable.add(repository.movies(movieId, list, page)
                .doOnSubscribe { view.loading(true) }
                .doAfterTerminate { view.loading(false) }
                .subscribe({
                    if (it.isEmpty()) {
                        view.error(EmptyViewMode.MODE_NO_MOVIES)
                        return@subscribe
                    }
                    view.content(it)
                }, { view.error(EmptyViewMode.MODE_NO_MOVIES) }))
    }

    override fun moviesNext(movieId: Int, list: String) {
        if (NetworkUtil.isNetworkConnected().not()) return

        page++
        disposable.add(repository.movies(movieId, list, page).subscribe { view.content(it) })
    }

    override fun movies(keywordId: Int) {
        if (NetworkUtil.isNetworkConnected().not()) {
            view.error(EmptyViewMode.MODE_NO_CONNECTION)
            return
        }

        page = 1
        disposable.add(repository.movies(keywordId, page)
                .doOnSubscribe { view.loading(true) }
                .doAfterTerminate { view.loading(false) }
                .switchIfEmpty { view.error(EmptyViewMode.MODE_NO_MOVIES) }
                .subscribe({ view.content(it) }, { view.error(EmptyViewMode.MODE_NO_MOVIES) }))
    }

    override fun moviesNext(keywordId: Int) {
        if (NetworkUtil.isNetworkConnected().not()) {
            return
        }

        page++
        disposable.add(repository.movies(keywordId, page).subscribe { view.content(it) })
    }

    override fun moviesWatchlist(accountId: Int, sessionId: String) {
        /*if (!NetworkUtil.isNetworkConnected()) {
            view?.error(EmptyViewMode.MODE_NO_CONNECTION)
            return
        }*/

        page = 1
        disposable.add(repository.moviesWatchlist(accountId, sessionId, page)
                .doOnSubscribe { view.loading(true) }
                .doAfterTerminate { view.loading(false) }
                .subscribe({
                    if (it.isEmpty()) {
                        view.error(1)
                        return@subscribe
                    }
                    view.content(it) }, { view.error((it as HttpException).code()) }
                )
        )
    }

    override fun moviesWatchlistNext(accountId: Int, sessionId: String) {
        if (!NetworkUtil.isNetworkConnected()) return

        page++
        disposable.add(repository.moviesWatchlist(accountId, sessionId, page)
                .subscribe { view.content(it) })
    }

    override fun moviesFavorite(accountId: Int, sessionId: String) {
        /*if (NetworkUtil.isNetworkConnected().not()) {
            view?.error(Throwable("no connection"))
            return
        }*/

        page = 1
        disposable.add(repository.moviesFavorite(accountId, sessionId, page)
                .doOnSubscribe { view.loading(true) }
                .doAfterTerminate { view.loading(false) }
                .subscribe(
                        {
                            if (it.isEmpty()) {
                                view.error(1)
                                return@subscribe
                            }
                            view.content(it) },
                        { view.error((it as HttpException).code())}
                )
        )
    }

    override fun moviesFavoriteNext(accountId: Int, sessionId: String) {
        if (NetworkUtil.isNetworkConnected().not()) {
            return
        }

        page++
        disposable.add(repository.moviesFavorite(accountId, sessionId, page)
                .subscribe { view.content(it)})
    }

    override fun destroy() {
        disposable.dispose()
    }
}